modules/fedimint-ln-client-ng/src/lightning_send.rs

```diff
@@ -0,0 +1,466 @@
+use std::sync::Arc;
+use std::time::Duration;
+
+use bitcoin_hashes::sha256;
+use fedimint_client::sm::{ClientSMDatabaseTransaction, State, StateTransition};
+use fedimint_client::transaction::ClientInput;
+use fedimint_client::DynGlobalClientContext;
+use fedimint_core::config::FederationId;
+use fedimint_core::core::OperationId;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::task::sleep;
+use fedimint_core::util::SafeUrl;
+use fedimint_core::{OutPoint, TransactionId};
+use fedimint_ln_common_ng::api::LnFederationApi;
+use fedimint_ln_common_ng::contracts::{ContractKey, HashTimeLockedContract};
+use fedimint_ln_common_ng::{
+    LightningClientContext, LightningInput, LightningWitness, OutgoingContractWitness,
+};
+use lightning_invoice::Bolt11Invoice;
+use secp256k1::schnorr::Signature;
+use secp256k1::KeyPair;
+use serde::{Deserialize, Serialize};
+use tracing::error;
+
+use crate::LightningClientStateMachines;
+
+const RETRY_DELAY: Duration = Duration::from_secs(1);
+const EXPIRATION_RETRY_DELAY: Duration = Duration::from_secs(10 * 60);
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// State machine of the gateway that relays a Lightning payment.
+///
+/// ```mermaid
+/// graph LR
+/// classDef virtual fill:#fff,stroke-dasharray: 5 5
+///
+///  Created -- funding tx is rejected --> FundingRejected
+///  Created -- funding tx is accepted --> Funded    
+///  Funded -- post invoice returns preimage  --> Success
+///  Funded -- post invoice returns forfeit tx --> Refunding
+///  Funded -- await_preimage returns preimage --> Success
+///  Funded -- timelock expires --> Refunding
+///  Funded -- post invoice returns invalid response --> GatewayError
```

Comment:
Do we want to sit in this temporary `GatewayError` state when the gateway does something unexpected? The scenario I'm thinking of is this: say there's a bug on the gateway side that causes an error to be returned and the client to transition to `GatewayError`. We fix the bug, then the gateway operator updates their gateway code and now the payment should succeed. In this case the client will never try again and will wait until the timelock....but they could try again and since they gateway has been updated, it should succeed.

Reply:
The alternative here would be to stay in the Funded state and retry like we would do in case of a network error. I have considered this and am actually still torn about it so good catch! It would make the state machine simpler. The small downside would be continuous retries - however we could do an exponential backoff here? Its a somewhat theoretical state since the gateway would have to return a invalid signature or the wrong preimage to trigger the transition to GatewayError.

Reply:
So I think it makes sense to just retry and remove the gateway error state. Just pushed a commit.

Reply:
+1 agreed with that. The retry delay in this scenario can be much longer than the network error retry